relay-server/src/actors/project_upstream.rs

```diff
@@ -1,25 +1,49 @@
 use std::borrow::Cow;
-use std::collections::{BTreeMap, HashMap};
+use std::collections::{hash_map::Entry, BTreeMap, HashMap};
 use std::sync::Arc;
-use std::time::{Duration, Instant};
+use std::time::Duration;
 
-use actix::fut;
-use actix::prelude::*;
+// TODO(actix): These two import will be removed when the `Upstream` actor is migrated to new tokio
+// runtime.
+use actix::SystemService;
 use actix_web::http::Method;
-use futures01::{future, future::Shared, sync::oneshot, Future};
+use futures::{future, FutureExt};
 use itertools::Itertools;
 use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc;
+use tokio::time::Instant;
 
 use relay_common::{ProjectKey, RetryBackoff};
 use relay_config::Config;
 use relay_log::LogError;
 use relay_statsd::metric;
+use relay_system::{
+    compat, BroadcastChannel, BroadcastResponse, BroadcastSender, FromMessage, Interface, Service,
+};
 
 use crate::actors::project::ProjectState;
 use crate::actors::project_cache::{FetchProjectState, ProjectError};
-use crate::actors::upstream::{RequestPriority, SendQuery, UpstreamQuery, UpstreamRelay};
+use crate::actors::upstream::{
+    RequestPriority, SendQuery, UpstreamQuery, UpstreamRelay, UpstreamRequestError,
+};
 use crate::statsd::{RelayCounters, RelayHistograms, RelayTimers};
-use crate::utils::{self, ErrorBoundary};
+use crate::utils::ErrorBoundary;
+
+/// This is the return type from [`UpstreamProjectSourceService::fetch_states`] function. Returns
+/// list of the results:
+/// * when the response successful, we get back the tuple of [`ProjectStateChannel`] tree and the
+///   results from the requests to upstream
+/// * otherwise we get the tuple of [`ProjectStateChannel`]
+///   tree which we can re-used to re-schedule the fetch or the channel error.
+type ProjectStatesResponses = Vec<
+    Result<
+        (
+            BTreeMap<ProjectKey, ProjectStateChannel>,
+            Result<GetProjectStatesResponse, UpstreamRequestError>,
+        ),
+        (BTreeMap<ProjectKey, ProjectStateChannel>, ProjectError),
+    >,
```

Comment:
Two comments on this, if possible:

1. Use `HashMap` here instead of `BTreeMap`. You could even create a typedef for that map, since this is used in several places.
2. It should be possible to flatten this type (and further the channel's value type) into a single structure with two fields: the result, and remaining channels.

Reply:
That's a very good point , done in [462a093](https://github.com/getsentry/relay/pull/1727/commits/462a09371fb1cfb026641ae515672790b4b1c593)